src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs

```diff
@@ -456,7 +456,40 @@ public static string GetPlatformApiLevel (string platformApiLevelName)
 
 		public static string GetLibraryImportDirectoryNameForAssembly (string assemblyIdentName)
 		{
-			return string.Concat (new MD2Managed ().ComputeHash (Encoding.UTF8.GetBytes (assemblyIdentName)).Take (5).Select (b => b.ToString ("X02")));
+			if (assemblyIdentityMap == null) {
+				throw new Exception ("Mapping was not loaded!");
+			}
+			if (assemblyIdentityMap.ContainsKey (assemblyIdentName)) {
+				return assemblyIdentityMap [assemblyIdentName];
+			}
+			string id = string.Format ("{0}", assemblyIdentityMap.Count + 1);
+			assemblyIdentityMap.Add (assemblyIdentName, id);
+			return id;
+		}
+
+		static Dictionary<string, string> assemblyIdentityMap;
+
+		public static void LoadAssemblyIdentityMap (string mapPath)
+		{
+			assemblyIdentityMap = new Dictionary<string, string> ();
+			if (!File.Exists (mapPath))
+				return;
+			foreach (var s in File.ReadLines (mapPath)) {
+				var items = s.Split ('=');
```

Comment:
Question: What should happen if the assembly contains `=`? `=` isn't a restricted character; an assembly named `foo=bar=baz.dll` is perfectly kosher.

What *will* happen is that things will *break* two lines down, because the assumption is that there is only one `=` on a line. If there are more than one, the first item is the assembly, the second is the value, but we'd *actually* have a line containing:

    foo=bar=baz=1

but then do `assemblyIdentityMap.Add("foo", "bar")`, which is wrong.

What we can *instead* do is reorder the values: instead of `ASSEMBLY=DIRECTORY`, do `DIRECTOR=ASSEMBLY`, then use the `string.Split(char[], int)` overload.

For example, if we instead had a line of:

    1=foo=bar=baz.dll

Then `line.Split(new[]{'='}, 2)` will return `new[]{ "1", "foo=bar=baz.dll" }`, which is fine.

This implies that the file format needs "reversing".